pkg/querier/multi_tenant_querier.go

```diff
@@ -1,15 +1,132 @@
 package querier
 
 import (
+	"context"
+
 	"github.com/go-kit/log"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/weaveworks/common/user"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/tenant"
+)
+
+const (
+	defaultTenantLabel   = "__tenant_id__"
+	retainExistingPrefix = "original_"
 )
 
 // MultiTenantQuerier is able to query across different tenants.
 type MultiTenantQuerier struct {
-	SingleTenantQuerier
+	Querier
+	resolver tenant.Resolver
 }
 
 // NewMultiTenantQuerier returns a new querier able to query across different tenants.
-func NewMultiTenantQuerier(querier SingleTenantQuerier, logger log.Logger) *MultiTenantQuerier {
-	return &MultiTenantQuerier{SingleTenantQuerier: querier}
+func NewMultiTenantQuerier(querier Querier, logger log.Logger) *MultiTenantQuerier {
+	return &MultiTenantQuerier{
+		Querier:  querier,
+		resolver: tenant.NewMultiResolver(),
+	}
+}
+
+func (q *MultiTenantQuerier) SelectLogs(ctx context.Context, params logql.SelectLogParams) (iter.EntryIterator, error) {
+
+	tenantIDs, err := q.resolver.TenantIDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(tenantIDs) == 1 {
+		singleContext := user.InjectUserID(ctx, tenantIDs[0])
+		return q.Querier.SelectLogs(singleContext, params)
+	}
+
+	iters := make([]iter.EntryIterator, len(tenantIDs))
+	for i, id := range tenantIDs {
+		singleContext := user.InjectUserID(ctx, id)
+		iter, err := q.Querier.SelectLogs(singleContext, params)
+
+		if err != nil {
+			return nil, err
+		}
+
+		iters[i] = NewTenantEntryIterator(id, iter)
+	}
+	return iter.NewSortEntryIterator(iters, params.Direction), nil
```

Comment:
I don't think we should deduplicate. What do you think, @cyriltovena

Reply:
No that's good !